crates/ibc/src/core/ics02_client/msgs/update_client.rs

```diff
@@ -3,32 +3,36 @@
 use crate::prelude::*;
 
 use ibc_proto::google::protobuf::Any;
+use ibc_proto::ibc::core::client::v1::MsgSubmitMisbehaviour as RawMsgSubmitMisbehaviour;
 use ibc_proto::ibc::core::client::v1::MsgUpdateClient as RawMsgUpdateClient;
 use ibc_proto::protobuf::Protobuf;
 
 use crate::core::ics02_client::error::ClientError;
 use crate::core::ics24_host::identifier::ClientId;
 use crate::signer::Signer;
-use crate::tx_msg::Msg;
 
-pub const TYPE_URL: &str = "/ibc.core.client.v1.MsgUpdateClient";
+pub const UPDATE_CLIENT_TYPE_URL: &str = "/ibc.core.client.v1.MsgUpdateClient";
+pub const MISBEHAVIOUR_TYPE_URL: &str = "/ibc.core.client.v1.MsgSubmitMisbehaviour";
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum UpdateClientKind {
+    /// this is the typical scenario where a new header is submitted to the client
+    /// to update the client
+    UpdateHeader,
+    /// this is the scenario where misbehaviour is submitted to the client
+    /// (e.g 2 headers with the same height in Tendermint)
+    Misbehaviour,
+}
```

Comment:
To my understanding, including `UpdateClientKind` here in ICS02 and its usage as an argument in some of the `ClientState` interfaces, requires it be client agnostic, flexible enough. So, users be able to add/remove their desired variants. This way, we implicitly assume the correct variants of a `client_message` while it is mainly adapted for the CometBFT clients. We neither have such a type within spec nor similar implementation in IBC-go which make me to not be sure about it.

- How about taking a closer look, and consider moving away from using the `UpdateClientKind` enum and `update_kind` fields? Couldn’t we add a `decode_client_message` method for the `ClientState` trait and calling it wherever needed? I am suggesting this for the same reason we added a `decode_client_state` API under the `ValidationContext` (The design required us!)

- A prerequisite for doing so is to merge [`misbehavior`](https://github.com/cosmos/ibc-rs/blob/a7231167efd2edb8fd44aadf4b23583b0158f27e/crates/ibc/src/core/ics02_client/misbehaviour.rs#L9) and [`Header`](https://github.com/cosmos/ibc-rs/blob/a7231167efd2edb8fd44aadf4b23583b0158f27e/crates/ibc/src/core/ics02_client/header.rs#L19) traits and define a new `ClientMessage` interface. Something similar to what already live in [IBC-go](https://github.com/cosmos/ibc-go/blob/d840c699d988e32cb55f2c289d69f8ef2e5cea50/modules/core/exported/client.go#L153-L161). This specific merge anyhow looks necessary and should be done along with refactoring others (See [here](https://github.com/cosmos/ibc-go/blob/main/docs/architecture/adr-006-02-client-refactor.md#merge-headermisbehaviour-interface-and-rename-to-clientmessage)).
    
- Also, this way, looking at the [`misbehavior.ValidateBasic()`](https://github.com/cosmos/ibc-go/blob/5a67efc496125823f6ca998048ce6c95dcc98427/modules/light-clients/07-tendermint/misbehaviour.go#L46) and [`header.ValidateBasic()`](https://github.com/cosmos/ibc-go/blob/143e5d85e9d2fc5fc2fc53adc42127a439ee2b79/modules/light-clients/07-tendermint/header.go#L50) in IBC-go, some of `verify_client_message` checks could be moved under a `validate_basic` method.

Reply:
> including UpdateClientKind here in ICS02 and its usage as an argument in some of the ClientState interfaces, requires it be client agnostic, flexible enough. So, users be able to add/remove their desired variants.

There is no flexibility to be had here. This is an internal type, meant to hold either a `Header` or a `Misbehaviour` message (where the specific "runtime type" depends on the light client); nothing else. The only reason we group them together is that we process both messages in the same way in [`update_client`](https://github.com/cosmos/ibc-rs/blob/b10f4baea722f3942f81fdeb610c84ad4e664f92/crates/ibc/src/core/ics02_client/handler/update_client.rs#L14-L85); this applies to only these 2 cases.

> Couldn’t we add a decode_client_message method for the ClientState trait and calling it wherever needed? I am suggesting this for the same reason we added a decode_client_state API under the ValidationContext (The design required us!)

This is different. We need `decode_client_state()` because by design, we don't want to have to know about all concrete `ClientState`s that exist. So when the relayer gives us a concrete `ClientState` in `CreateClient`, we don't know how to decode it *by design*. We rely on the host for that, since the host knows about all the light clients that it supports.

Here, *we create the domain `MsgUpdateClient`*. The `update_kind` field is just so that we remember what "client message" *we put in there*; and we will only ever put one of 2 things: an `UpdateHeader` or a `Misbehaviour`.

Reply:
changed some names and improved the docstrings, see if you like it [ea2d588](https://github.com/cosmos/ibc-rs/pull/584/commits/ea2d588b84ef630858aea948920d3d758e1665d8)

Reply:
and I forgot we had the `Misbehavior` trait; it was not used, so I just removed it [af10978](https://github.com/cosmos/ibc-rs/pull/584/commits/af109786f551da630fb3b6ae9a03944e064c9a7f)

Reply:
> changed some names and improved the docstrings, see if you like it [ea2d588](https://github.com/cosmos/ibc-rs/pull/584/commits/ea2d588b84ef630858aea948920d3d758e1665d8)

Better now 🙏🏻 

> and I forgot we had the Misbehavior trait; it was not used, so I just removed it [af10978](https://github.com/cosmos/ibc-rs/pull/584/commits/af109786f551da630fb3b6ae9a03944e064c9a7f)

I couldn't find any usage for this [`Header`](https://github.com/cosmos/ibc-rs/blob/b10f4baea722f3942f81fdeb610c84ad4e664f92/crates/ibc/src/clients/ics07_tendermint/header.rs#L84-L92)  implementation, too
And, basically, It's not clear what's the [`Header`](https://github.com/cosmos/ibc-rs/blob/ef52941e29aa2759e2861cdae8a7398b114cc5f2/crates/ibc/src/core/ics02_client/header.rs#L18-L42) trait role. Only serving the `mock` tests

Reply:
Created an issue for it as it is also used in `RelayerContext`, which is currently used in our mock code, which is not test-only. I'd like to investigate moving the whole mock code to `tests`, and not exporting it anymore, as it's not clear what value it currently offers, and couples our test code with non-test code in annoying ways.

https://github.com/cosmos/ibc-rs/issues/617